### PR TITLE
Support for decoding jpegs on GPU with nvjpeg

### LINKED ISSUE #2742: Implement JPEG decoding via nvjpeg
## 🚀 Feature

We currently provide CPU image decoding functions for PNG and JPEG. It could be interesting to consider using nvjpeg to perform decoding directly on the GPU, as it could benefit certain workloads.

## Motivation

With CUDA 10.0, nvjpeg is [part of the CUDA toolkit](https://developer.nvidia.com/nvjpeg), which means that it would be easily accessible for us to use it.

The API for the decoding function is not yet clear to me, and would require some digging
Ideally, passing a CUDA tensor for `decode_jpeg` should do the decoding on the GPU, although I'm not sure if that's how nvjpeg works internally.

### PR DESCRIPTION
**Pull request by NicolasHug**
Closes https://github.com/pytorch/vision/pull/2786/
Closes #2742

This is based on @jamt9000's great initial work in https://github.com/pytorch/vision/pull/2786/. I mostly made some minor clean ups and added some tests.

In terms of usage, the current supported API is to add a `device` parameter to `io.image.decode_jpeg`.

For benchmarks, see https://github.com/pytorch/vision/pull/2786#issuecomment-832148710. Overall this seems to offer a 2-3X speedup over CPU decoding (without libjpeg-turbo).

Note: We use `nvjpegCreateSimple()` which is only available in cuda >= 10.1. So even though nvjpeg exists for 10.0, this won't compile. I assume this is OK since our CI only tests 10.1 and upwards anyway.

While we can move forward with this simple basic version, there seem to be room for improvement. In particular:
- Support for A100 devices
- Support for batch decoding (I didn't see any improvement in my experiments in https://github.com/pytorch/vision/pull/2786#issuecomment-832148710, but perhaps I missed something)
- Use a finer-grained API for the decoding phases, and potentially change the decoding backend depending on the image size, taking inspiration from https://github.com/NVIDIA/CUDALibrarySamples/tree/master/nvJPEG/nvJPEG-Decoder-MultipleInstances

### CODE REVIEW

**Review by datumbox — COMMENTED** [review]
Thanks for the PR! 

I'll have a more thorough look next week. I only had a quick check and left a few comments for discussion. Let me know your thoughts.

Edit: I had a second look. Overall the PR looks great. I flagged two more things to discuss prior merging. With the exception of the potential memory leak that we need to investigate, all other comments are questions around the API so there is no need to modify your code.

File: torchvision/csrc/io/image/cuda/decode_jpeg_cuda.cpp
Comment (NicolasHug) [reply torchvision/csrc/io/image/cuda/decode_jpeg_cuda.cpp:28]: > Also just checking whether this should be released at any point to avoid memory leaks

Yeah this is a good point. Creating / allocating it at each call has some severe overhead so it makes sense to declare this as a global variable (related discussion: https://github.com/pytorch/vision/pull/2786#discussion_r564131673). But this means we never really know when to release it, and the memory will only be freed when the process is killed.

File: test/test_image.py
Comment (datumbox) [inline test/test_image.py:313]: What is the reason for requiring the input to be on CPU? Is there a specific limitation on the API of the library? 

If it is possible to have the data on CUDA, then it would be able to call `decode_jpeg()` without specifying a device and assume that the data should remain where they are. This has the benefit of making your methods work exactly like any other operator in TorchVision and potentially might allow you to use the dispatcher (needs to be confirmed).

File: torchvision/io/image.py
Comment (datumbox) [inline torchvision/io/image.py:152]: Just flagging that there are other methods such as the `read_image()` and `decode_image()` to which you can also expose the GPU support. If you choose not to introduce the `device` argument to them, they will only decompress JPEG on CPU. Personally, I agree with your approach and I would also keep the feature contained for now. Once it becomes stable it would be very easy to expose it via the other APIs in separate PRs.

File: test/test_image.py
Comment (jamt9000) [reply test/test_image.py:313]: The first stage of nvjpeg decoding is done on the CPU, so it needs to take a CPU buffer as input.

Comment (fmassa) [reply test/test_image.py:313]: Another reason I believe why this takes the input tensor on a CPU device is because we need to get some metadata from the Tensor to allocate the output tensor, which would mean doing a CUDA->CPU copy to perform the allocation.

File: torchvision/csrc/io/image/cuda/decode_jpeg_cuda.cpp
Comment (fmassa) [inline torchvision/csrc/io/image/cuda/decode_jpeg_cuda.cpp:55]: I'm wondering if we should clear this if the creation fails, otherwise we might not be able to run this function anymore as the handle will be invalid?

**Review by fmassa — APPROVED** [review]
Looks great to me, thanks!

I've made a few comments, I none of which are merge-blocking I think.

File: torchvision/csrc/io/image/image.cpp
Comment (datumbox) [reply torchvision/csrc/io/image/image.cpp:25]: Thanks for the clarifications concerning nvjpeg. I think that we can investigate on future PRs how we could do this more elegantly. No need to block this PR.

File: torchvision/csrc/io/image/cuda/decode_jpeg_cuda.cpp
Comment (datumbox) [reply torchvision/csrc/io/image/cuda/decode_jpeg_cuda.cpp:49]: Good spot. I think that's the only issue that we should address before merging.

File: test/test_image.py
Comment (NicolasHug) [reply test/test_image.py:296]: The max can be quite high unfortunately, the minimum seems to be 52, after which some tests start failing.
In `test_decode_jpeg`, we also test for the MAE (with the same threshold=2)

File: torchvision/csrc/io/image/cuda/decode_jpeg_cuda.cpp
Comment (datumbox) [inline torchvision/csrc/io/image/cuda/decode_jpeg_cuda.cpp:51]: non-blocking nit: Might be worth doing double-checked locking here to speed it up.

**Review by datumbox — APPROVED** [review]
LGTM!

**Comment by NicolasHug** [comment]
CI is green(ish) so I'll merge.

Thanks everyone for the reviews and especially @jamt9000 for the initial work!
I'll follow up and open an issue with potential future improvements

File: torchvision/csrc/io/image/cuda/decode_jpeg_cuda.cpp
Comment (jamt9000) [inline torchvision/csrc/io/image/cuda/decode_jpeg_cuda.cpp:59]: Since it's an opaque handle I think the use of `free()` may not be correct unless it's documented as being supported.

(I would hope that it simply leaves the handle as null if initialisation fails, although I don't see that in the docs - [here it just reinits the handle without any freeing when hw backend fails though](https://github.com/NVIDIA/CUDALibrarySamples/blob/eb89d592ee4c05a9b358cdc1888e29bb00760d99/nvJPEG/nvJPEG-Decoder/nvjpegDecoder.cpp#L334))

Comment (NicolasHug) [reply torchvision/csrc/io/image/cuda/decode_jpeg_cuda.cpp:59]: Hm that's a good point. What would you recommend instead of `free`?

Before pushing this I did a quick test by inserting

```C++
       nvjpegStatus_t create_status = nvjpegCreateSimple(&nvjpeg_handle);
       create_status = NVJPEG_STATUS_NOT_INITIALIZED;   // <- this
```

and all the tests  were failing gracefully  with `E   RuntimeError: nvjpegCreateSimple failed: 1`. Since I was running the tests with `pytest test/test_image.py -k cuda` they were all in the same process and `pytest` was just catching the `RuntimeErrors`, so I assumed it was OK.

File: test/test_image.py
Comment (fmassa) [reply test/test_image.py:296]: Hum, this looks suspicious that the decoding gives such large differences. Something to keep an eye on

**Comment by ShiyangZhang** [comment]
Hi guys,
from here https://github.com/pytorch/vision/blob/82f9a187681dade1620bc06b85f317a11aea40dc/torchvision/csrc/io/image/cuda/decode_jpeg_cuda.cpp#L74

Why  nvjpeg_handle_creation_flag is  local ?

**Pull request merged**
